Feature: Add UPE availability inbox notice

Mayisha opened the pull request:
Fixes #1692

### Description
Adding an inbox notice for the merchants informing them about the availability of the new payment experience in Stripe.

![inbox](https://user-images.githubusercontent.com/33387139/131287918-39b68575-89d4-477d-a465-79b9eb36ef3a.png)

### Testing instructions
- To make the notice visible in the inbox 
-> Connect Stripe in your store to make the notice available
-> Enable UPE preview flag (returning `true` from [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/24277919aa7b9029c39ecd997dd133990e7f65c7/includes/class-wc-stripe-feature-flags.php#L15) should do the work)
-> UPE should be disabled (returning `false` from [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/24277919aa7b9029c39ecd997dd133990e7f65c7/includes/class-wc-stripe-feature-flags.php#L24) should do the work)
- To check the conditions when notice should not be visible 
-> delete the already created notice from DB (`wp_wc_admin_notes` and `wp_wc_admin_note_actions` table) 
-> Make sure the conditions mentioned above are unsupported
-> Refresh the page

### QA checklist

- [ ] When Stripe is connected, UPE is disabled and UPE preview flag is enabled a new inbox notice titled `Boost your sales with the new payment experience in Stripe` appears.
- [ ] The notice does not appear if the conditions are unsupported.

Mayisha committed:
Add inbox notice about UPE availability

Mayisha committed:
Add changelog entry

Mayisha committed:
Fix conditions to show UPE availability inbox notice

harriswong commented:
👀

harriswong commented on includes/notes/class-wc-stripe-upe-availability-note.php:
```diff
@@ -0,0 +1,70 @@
+	/**
+	 * Link to enable the UPE in store.
+	 */
+	const ENABLE_IN_STORE_LINK = '?TODO';
```
We added a blank page for the onboarding page [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1737). You can use `wp-admin/admin.php?page=wc_stripe-onboarding_wizard` here.

harriswong reviewed (commented):
Tested the note and confirmed it shows in inbox with conditions 👍 
- stripe is not enable/disabled
- `WC_Stripe_Feature_Flags::is_upe_checkout_enabled()` true/false
- `WC_Stripe_Feature_Flags::is_upe_settings_redesign_enabled()` true/false

What do you think about adding a test for the note? Either unit test for `WC_Stripe_UPE_Availability_Note` to test the note itself, or a test for `WC_Stripe_Inbox_Notes::create_upe_availability_note()` to confirm the note it generates?

Mayisha committed:
Update value of 'ENABLE_IN_STORE_LINK'

Mayisha commented:
@harriswong Thank you for your feedback. I have updated the PR with the following changes-
- Updated the `Enable in your store` link to `?page=wc_stripe-onboarding_wizard`
- Added test for the note itself to test different properties of the note
- Added test for `WC_Stripe_Inbox_Notes::create_upe_availability_note()` to test note creation on different conditions.

Mayisha committed:
Add tests for UPE availability note

Mayisha committed:
Add compatibility check

Mayisha commented:
@harriswong fixed the compatibility issues. :smiley:

harriswong commented on includes/admin/class-wc-stripe-inbox-notes.php:
```diff
@@ -31,6 +32,24 @@ public function __construct() {
+			return;
+		}
+
+		require_once WC_STRIPE_PLUGIN_PATH . '/includes/notes/class-wc-stripe-upe-availability-note.php';
```
Do we need this line? If the class `WC_Stripe_UPE_Availability_Note` does not exist, it would have exited in line 45-46. So then if it goes through the condition there, this file is already loaded?

harriswong commented on woocommerce-gateway-stripe.php:
```diff
@@ -154,6 +154,7 @@ public function init() {
 				require_once dirname( __FILE__ ) . '/includes/compat/class-wc-stripe-pre-orders-compat.php';
 				require_once dirname( __FILE__ ) . '/includes/class-wc-gateway-stripe.php';
 				require_once dirname( __FILE__ ) . '/includes/class-wc-stripe-feature-flags.php';
+				require_once dirname( __FILE__ ) . '/includes/notes/class-wc-stripe-upe-availability-note.php';
```
Instead of loading the Note class and then checks `trait_exists()` inside, what if we do something similar to WCPay where we check it here? https://github.com/Automattic/woocommerce-payments/blob/d4e86120cacdc077cd4858ea7e9a9071358b6630/includes/class-wc-payments.php#L878-L893

```
if (version > 4.4 ) {
  require_once dirname( __FILE__ ) . '/includes/notes/class-wc-stripe-upe-availability-note.php';
}
```

Would this help preventing the entire note to load without adding all the `if ( class_exists( 'WC_Stripe_UPE_Availability_Note',` everywhere?

Mayisha replied:
This actually solved the issue. I changed the implementation mention in [this](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1762#discussion_r702152140) and [this](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1762#pullrequestreview-746460458) comments also after doing this and everything seems to be in place now.
Thanks for pointing out in the right direction :grinning:

harriswong reviewed (commented):
Tested and I can see the note in inbox. I am not too sure about wrapping `if ( ! class_exists( 'WC_Stripe_UPE_Availability_Note', false )` around test cases and wrapping an entire class with `if ( trait_exists( 'Automattic\WooCommerce\Admin\Notes\NoteTraits' ) )`, but if this is the only way to go to pass the 3.3.0 compat tests then 👍

Mayisha committed:
Address feedbacks

frosso commented on includes/notes/class-wc-stripe-upe-availability-note.php:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	public static function init() {
+		if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '4.4.0', '>=' ) ) {
+			self::possibly_add_note();
+		}
+	}
```
This version check exists because older versions of WC don't include a `NoteTraits` trait.

So, if this file is included on older versions of WC, there would be a fatal error saying that `NoteTraits` doesn't exist, even before this check is executed.

I'd recommend placing this check _before_ the inclusion of the file (and instantiation of the class), and maybe moving the checks in `create_upe_availability_note` into this `init` static method.

Mayisha replied:
Thanks for the suggestions. 

The check is already present before the inclusion of the file [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2251ac7f417806fbe73843bf8787dfca1bc73152/woocommerce-gateway-stripe.php#L214). (This issue with `NoteTraits` was caught in the compatibility check for version 3.0.0 and then fixed it in a later commit)

I have addressed the second feedback about moving the checks inside the init method in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1762/commits/3f9bdca1e2a86d1b19b6ecf5fcfbc944e6860a39). Also removed the unnecessary duplicate check for version as it is already checked where the file is included.

Mayisha committed:
Move conditional checks inside the init method

frosso commented on includes/admin/class-wc-stripe-inbox-notes.php:
```diff
@@ -31,6 +32,10 @@ public function __construct() {
 	}
 
+	public static function create_upe_availability_note() {
+		WC_Stripe_UPE_Availability_Note::init();
```
The compatibility issue would still persist, because on `WC_VERSION` with value < than `4.4.0` the `WC_Stripe_UPE_Availability_Note` class would not exist (because its file is not included). So this line would cause a fatal error.
```suggestion
        if ( version_compare( WC_VERSION, '4.4.0', '>=' ) ) {
            require_once WC_STRIPE_PLUGIN_PATH . '/includes/notes/class-wc-stripe-upe-availability-note.php';
    		WC_Stripe_UPE_Availability_Note::init();
		}
```

Mayisha committed:
check condition before init

frosso commented on changelog.txt:
```diff
@@ -7,6 +7,7 @@
 * Add - Settings to control Payment Request Button locations in the Stripe plugin settings. Persists changes made through pre-existing filters, or defaults to the Cart and Product pages if no filters are in use.
 * Tweak - Deprecated the 'wc_stripe_hide_payment_request_on_product_page', 'wc_stripe_show_payment_request_on_checkout', and 'wc_stripe_show_payment_request_on_cart' filters in favor of the UI-driven approach in the plugin settings.
 * Add - Notice for WP & WC version compatibility check.
+* Add - Notice about UPE availability.
```
(minor) Is a changelog entry necessary in this case? 🤔 
I'm thinking we can have one changelog entry to notify merchants of UPE, without having a specific one just for the notice.

harriswong replied:
i wondered about the same thing, but +1 on having just 1 for the entire Stripe UPE feature. This is similar to what we did for WCPay.

frosso reviewed (approved):
LGTM 👍

harriswong reviewed (approved):
- Tested and the notes appears in inbox
- clicked "enable in your store" and it leads me to onboarding wizard 👍 
- Not sure if we need to remove the note after it's clicked. Or, if we should remove the note when the user enabled UPE. But this could be a follow-up cc @frosso 

👍 🚢

Mayisha committed:
Remove changelog entry
